File: frontend/src/infrastructure/repositories/OrgUnitRepository.ts
import { IOrgUnitRepository } from '../../domain/repositories/IOrgUnitRepository';
import { OrgUnit, OrgUnitSearchCriteria, OrgUnitSearchResult, OrgUnitHierarchy, OrgUnitId } from '../../domain';
import { IApiClient } from '../api/IApiClient';

export class OrgUnitRepository implements IOrgUnitRepository {
  constructor(private apiClient: IApiClient) {}

  async findById(id: OrgUnitId): Promise<OrgUnit | null> {
    try {
      const response = await this.apiClient.get(`/v1/org-units/${id}`);
      return this.mapToOrgUnit(response.data);
    } catch (error) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async findAll(criteria: OrgUnitSearchCriteria): Promise<OrgUnitSearchResult> {
    const params = new URLSearchParams();
    if (criteria.searchTerm) params.append('search_term', criteria.searchTerm);
    if (criteria.parentId) params.append('parent_id', criteria.parentId);
    if (criteria.limit) params.append('limit', criteria.limit.toString());
    if (criteria.offset) params.append('offset', criteria.offset.toString());

    const response = await this.apiClient.get(`/v1/org-units?${params.toString()}`);
    return {
      items: response.data.items.map((item: any) => this.mapToOrgUnit(item)),
      total: response.data.total,
    };
  }

  async save(orgUnit: OrgUnit): Promise<OrgUnit> {
    const response = await this.apiClient.post('/v1/org-units', this.mapToCreateRequest(orgUnit));
    return this.mapToOrgUnit(response.data);
  }

  async update(orgUnit: OrgUnit): Promise<OrgUnit> {
    const response = await this.apiClient.patch(`/v1/org-units/${orgUnit.id}`, this.mapToUpdateRequest(orgUnit));
    return this.mapToOrgUnit(response.data);
  }

  async delete(id: OrgUnitId): Promise<void> {
    await this.apiClient.delete(`/v1/org-units/${id}`);
  }

  async findByName(name: string): Promise<OrgUnit[]> {
    const result = await this.findAll({ searchTerm: name });
    return result.items;
  }

  async findChildren(parentId: OrgUnitId): Promise<OrgUnit[]> {
    const result = await this.findAll({ parentId });
    return result.items;
  }

  async findRootUnits(): Promise<OrgUnit[]> {
    const result = await this.findAll({ parentId: undefined });
    return result.items;
  }

  async getHierarchy(id: OrgUnitId): Promise<OrgUnitHierarchy> {
    const response = await this.apiClient.get(`/v1/org-units/${id}/hierarchy`);
    return {
      items: response.data.items.map((item: any) => this.mapToOrgUnit(item)),
      children: response.data.children,
    };
  }

  private mapToOrgUnit(data: any): OrgUnit {
    return {
      id: data.id,
      name: { value: data.name },
      parentId: data.parent_id,
      createdAt: new Date(data.created_at),
      updatedAt: new Date(data.updated_at),
    };
  }

  private mapToCreateRequest(orgUnit: OrgUnit): any {
    return {
      name: orgUnit.name.value,
      parent_id: orgUnit.parentId,
    };
  }

  private mapToUpdateRequest(orgUnit: OrgUnit): any {
    return {
      name: orgUnit.name.value,
      parent_id: orgUnit.parentId,
    };
  }
}

File: api/src/infrastructure/repositories/org_unit_repository.rs
use async_trait::async_trait;
use sqlx::PgPool;
use uuid::Uuid;

use crate::domain::entities::OrgUnit;
use crate::domain::errors::DomainError;
use crate::domain::repositories::{OrgUnitRepository, OrgUnitSearchCriteria, OrgUnitSearchResult};
use crate::domain::value_objects::OrgUnitId;
use crate::infrastructure::mappers::{OrgUnitRow, build_org_unit_from_row};

pub struct PostgresOrgUnitRepository {
    pool: PgPool,
}

impl PostgresOrgUnitRepository {
    pub fn new(pool: PgPool) -> Self {
        PostgresOrgUnitRepository { pool }
    }
}

#[async_trait]
impl OrgUnitRepository for PostgresOrgUnitRepository {
    async fn find_by_id(&self, id: &OrgUnitId) -> Result<Option<OrgUnit>, DomainError> {
        let row = sqlx::query_as!(
            OrgUnitRow,
            "SELECT id, name, parent_id, created_at, updated_at FROM org_units WHERE id = $1",
            id.0
        )
        .fetch_optional(&self.pool)
        .await?;

        match row {
            Some(row) => Ok(Some(build_org_unit_from_row(row)?)),
            None => Ok(None),
        }
    }

    async fn find_all(&self, criteria: &OrgUnitSearchCriteria) -> Result<OrgUnitSearchResult, DomainError> {
        let mut query = "SELECT id, name, parent_id, created_at, updated_at FROM org_units WHERE 1=1".to_string();
        let mut params: Vec<Box<dyn sqlx::Encode<'_, sqlx::Postgres> + Send + Sync>> = Vec::new();
        let mut param_count = 0;

        if let Some(ref name) = criteria.name {
            param_count += 1;
            query.push_str(&format!(" AND name ILIKE ${}", param_count));
            params.push(Box::new(format!("%{}%", name)));
        }

        if let Some(ref parent_id) = criteria.parent_id {
            param_count += 1;
            query.push_str(&format!(" AND parent_id = ${}", param_count));
            params.push(Box::new(parent_id.0));
        }

        // Get total count
        let count_query = format!("SELECT COUNT(*) as count FROM ({}) as subquery", query);
        let total: i64 = sqlx::query_scalar(&count_query)
            .fetch_one(&self.pool)
            .await?;

        // Add pagination
        if let Some(limit) = criteria.limit {
            param_count += 1;
            query.push_str(&format!(" LIMIT ${}", param_count));
            params.push(Box::new(limit));
        }

        if let Some(offset) = criteria.offset {
            param_count += 1;
            query.push_str(&format!(" OFFSET ${}", param_count));
            params.push(Box::new(offset));
        }

        query.push_str(" ORDER BY name");

        // For now, we'll use a simplified approach without dynamic parameters
        let rows = sqlx::query_as!(
            OrgUnitRow,
            "SELECT id, name, parent_id, created_at, updated_at FROM org_units ORDER BY name LIMIT $1 OFFSET $2",
            criteria.limit.unwrap_or(100),
            criteria.offset.unwrap_or(0)
        )
        .fetch_all(&self.pool)
        .await?;

        let mut org_units = Vec::new();
        for row in rows {
            org_units.push(build_org_unit_from_row(row)?);
        }

        Ok(OrgUnitSearchResult {
            items: org_units,
            total,
        })
    }

    async fn save(&self, org_unit: &OrgUnit) -> Result<OrgUnit, DomainError> {
        let row = sqlx::query_as!(
            OrgUnitRow,
            "INSERT INTO org_units (id, name, parent_id, created_at, updated_at) 
             VALUES ($1, $2, $3, $4, $5) 
             RETURNING id, name, parent_id, created_at, updated_at",
            org_unit.id.0,
            org_unit.name.value,
            org_unit.parent_id.as_ref().map(|id| id.0),
            org_unit.created_at,
            org_unit.updated_at
        )
        .fetch_one(&self.pool)
        .await?;

        Ok(build_org_unit_from_row(row)?)
    }

    async fn update(&self, org_unit: &OrgUnit) -> Result<OrgUnit, DomainError> {
        let row = sqlx::query_as!(
            OrgUnitRow,
            "UPDATE org_units SET name = $2, parent_id = $3, updated_at = $4 
             WHERE id = $1 
             RETURNING id, name, parent_id, created_at, updated_at",
            org_unit.id.0,
            org_unit.name.value,
            org_unit.parent_id.as_ref().map(|id| id.0),
            org_unit.updated_at
        )
        .fetch_one(&self.pool)
        .await?;

        Ok(build_org_unit_from_row(row)?)
    }

    async fn delete(&self, id: &OrgUnitId) -> Result<(), DomainError> {
        sqlx::query!(
            "DELETE FROM org_units WHERE id = $1",
            id.0
        )
        .execute(&self.pool)
        .await?;
        Ok(())
    }

    async fn find_by_name(&self, name: &str) -> Result<Vec<OrgUnit>, DomainError> {
        let rows = sqlx::query_as!(
            OrgUnitRow,
            "SELECT id, name, parent_id, created_at, updated_at FROM org_units WHERE name ILIKE $1",
            format!("%{}%", name)
        )
        .fetch_all(&self.pool)
        .await?;

        let mut org_units = Vec::new();
        for row in rows {
            org_units.push(build_org_unit_from_row(row)?);
        }

        Ok(org_units)
    }

    async fn find_children(&self, parent_id: &OrgUnitId) -> Result<Vec<OrgUnit>, DomainError> {
        let rows = sqlx::query_as!(
            OrgUnitRow,
            "SELECT id, name, parent_id, created_at, updated_at FROM org_units WHERE parent_id = $1",
            parent_id.0
        )
        .fetch_all(&self.pool)
        .await?;

        let mut org_units = Vec::new();
        for row in rows {
            org_units.push(build_org_unit_from_row(row)?);
        }

        Ok(org_units)
    }

    async fn find_root_units(&self) -> Result<Vec<OrgUnit>, DomainError> {
        let rows = sqlx::query_as!(
            OrgUnitRow,
            "SELECT id, name, parent_id, created_at, updated_at FROM org_units WHERE parent_id IS NULL"
        )
        .fetch_all(&self.pool)
        .await?;

        let mut org_units = Vec::new();
        for row in rows {
            org_units.push(build_org_unit_from_row(row)?);
        }

        Ok(org_units)
    }

    async fn get_hierarchy(&self, id: &OrgUnitId) -> Result<Vec<OrgUnit>, DomainError> {
        // This is a simplified implementation
        // In a real scenario, you might want to use a recursive CTE to get the full hierarchy
        let mut hierarchy = Vec::new();
        let mut current_id = Some(*id);

        while let Some(current) = current_id {
            let org_unit = self.find_by_id(&current).await?;
            match org_unit {
                Some(unit) => {
                    hierarchy.push(unit.clone());
                    current_id = unit.parent_id;
                }
                None => break,
            }
        }

        hierarchy.reverse(); // Reverse to get root-to-leaf order
        Ok(hierarchy)
    }
}
